src/cli.rs

```diff
@@ -17,99 +17,73 @@ use argh::FromArgs;
-    /// list devices as full paths (not names)
-    #[argh(switch, short = 'P')]
-    pub paths: bool,
-}
-
-/// get the current brightness value
-#[derive(FromArgs, PartialEq, Debug)]
-#[argh(subcommand, name = "get")]
-pub struct ActionGet {
-    /// show the value as a percentage
-    #[argh(switch, short = 'p')]
-    pub percent: bool,
-    // /// percentage curve function (raw to percent)
-    // #[argh(option)]
-    // pub curve: Option<String>,
-}
-
-/// set the brightness value
-#[derive(FromArgs, PartialEq, Debug)]
-#[argh(subcommand, name = "set")]
-pub struct ActionSet {
-    /// percentage or value to set
-    #[argh(positional)]
-    pub value: Value,
-    /// only increase, never decrease
-    #[argh(switch, short = 'I')]
-    pub increase: bool,
-    /// only decrease, never increase
-    #[argh(switch, short = 'D')]
-    pub decrease: bool,
-    // /// percentage curve function
-    // #[argh(option)]
-    // pub curve: Option<String>,
-    /// duration of time when interpolating between 0% and 100%
-    #[argh(option, short = 't')]
-    pub duration: Option<DurationInterval>,
-}
-
-/// increase the brightness value
-#[derive(FromArgs, PartialEq, Debug)]
-#[argh(subcommand, name = "inc")]
-pub struct ActionIncrease {
-    /// percentage or value to add
-    #[argh(positional)]
-    pub amount: Value,
-    // /// percentage curve function
-    // #[argh(option)]
-    // pub curve: Option<String>,
-    /// duration of time over which to interpolate the change
-    #[argh(option, short = 't')]
-    pub duration: Option<DurationInterval>,
-}
-
-/// decrease the brightness value
-#[derive(FromArgs, PartialEq, Debug)]
-#[argh(subcommand, name = "dec")]
-pub struct ActionDecrease {
-    /// percentage or value to subtract
-    #[argh(positional)]
-    pub amount: Value,
-    // /// percentage curve function
-    // #[argh(option)]
-    // pub curve: Option<String>,
-    /// duration of time over which to interpolate the change
-    #[argh(option, short = 't')]
-    pub duration: Option<DurationInterval>,
+    /// list all discovered backlight devices
+    #[bpaf(command("list"))]
+    List {
+        /// list devices as full paths (not names)
+        #[bpaf(short('P'), long)]
+        paths: bool,
+    },
+    /// get the current brightness value
+    #[bpaf(command("get"))]
+    Get {
+        /// show the value as a percentage
+        #[bpaf(short('p'), long)]
+        percent: bool,
+    },
+    /// set the brightness value
+    #[bpaf(command("set"))]
+    Set {
+        /// percentage or value to set
+        #[bpaf(positional("VALUE"))]
+        value: Value,
+        /// only increase, never decrease
+        #[bpaf(short('I'), long("inc"))]
+        increase: bool,
+        /// only decrease, never increase
+        #[bpaf(short('D'), long("dec"))]
+        decrease: bool,
+        /// duration of time when interpolating between 0% and 100%
+        #[bpaf(short('t'), long)]
+        duration: Option<DurationInterval>,
+    },
```

Comment:
At the moment positional items must go at the end of the consumer. I'm slowly working on a more straightforward replacement, but it's still some time away.

```suggestion
    Set {
        /// only increase, never decrease
        #[bpaf(short('I'), long("inc"))]
        increase: bool,
        /// only decrease, never increase
        #[bpaf(short('D'), long("dec"))]
        decrease: bool,
        /// duration of time when interpolating between 0% and 100%
        #[bpaf(short('t'), long)]
        duration: Option<DurationInterval>,
        /// percentage or value to set
        #[bpaf(positional("VALUE"))]
        value: Value,
    },
```

You can also add a test to find issues like that using `check_invariants`:

https://docs.rs/bpaf/0.9.20/bpaf/struct.OptionParser.html#method.check_invariants